drafts/inlinemacros.md

```diff
@@ -165,99 +170,66 @@ positions:
-the original tree are kept after the rewrite. There is one exception to this: References to
-overloaded methods are re-computed after the rewrite. So a different alternative might
-be chosen if the argument types after the rewrite are more specific than the method
-parameter types. So, compile-time overloading resolution is essentially multi-method dispatch.
-Here is a scenario where this can be used to support constant folding inline infix methods.
-
-The task here is to define `++` as an infix operator for inline Ints, which should
-give an inline result if and only if its two operands are inline values. Here's how it can be achieved:
-
-    inline implicit class PlusWrapper(val x: Int) {
-      inline def ++ (y: Int): Int = plus(this.x, y)
-    }
-    inline def plus(x: inline Int, y: inline Int): inline Int = x + y
-    def plus(x: Int, y: Int): Int = x + y
-
-Let's assume we have
-
-    a, b: inline Int
-
-Then we rewrite `a ++ b` as follows:
-
-    a ++ b
-    -->
-    PlusWrapper(a).++(b)
-    -->
-    new PlusWrapper(a).++(b)
-    -->
-    plus(new PlusWrapper(a).x, b)
-    -->
-    plus(a, b)
-    -->
-    a + b: inline Int
-
-Note that the call to `plus` in `PlusWrapper` resolves statically to the non-inline method `plus`.
-Yet in the last-but-one expression of the above rewrite sequence it is rewritten to refer to the
-inline method `plus` because at that point both arguments are known to be inline values.
-
-
 ## Macros
 
-A macro is an expression of the form
+A macro expression is an expression of the form
 
     macro { ... }
 
-where `{ ... }` is some block of Scala code. (In fact, `macro` may prefix arbitrary expressions,
-but blocks are used most commonly). Macros may only appear in the bodies of inline methods.
-
-Macros can only reference the following things in their environment:
-
- 1. parameters of enclosing inline methods (where the parameter may, or may not be, marked inline),
- 2. inline values,
- 3. `this` references to classes containing an enclosing inline method as a member
- 4. anything that's global,
-
-A macro is said to _close over_ items in (1)-(3).
-
-Inside a macro, the type of closed-over definitions changes. If the definition is of type
-`inline T` outside the macro, it is of type `T` inside. If it is a non-inline type `U`, the type in
-the macro is `Expr[U]`, the type of expression trees of type `U`. Types with several levels of
-`inline` are handled analogously. For instance, a type like
-
-    inline (inline Int, Boolean) => String
-
-would be seen inside a macro as
-
-    (Int, Expr[Boolean]) => Expr[String]
-
-A dual transformation is applied to a macro's result. Normal types are prefixed with `inline` until
-a type is of the form `Expr[T]` in which case it is replaced by `T`. So if a macro body returned
-a list of type
-
-    List[(String, Expr[T])]
-
-the type of the macro itself would be
-
-    inline List[inline (inline String, T)]
```

Comment:
I believe that `Liftable` and `Unliftable` give a huge increase of what could be done in macros. It would be good to go with 1 way by default and include implementations for most of classes present in standard library.